Marlin/src/module/stepper.cpp

```diff
@@ -2862,18 +2892,185 @@ hal_timer_t Stepper::block_phase_isr() {
+            rate = block->cruise_rate - rate;
+            NOLESS(rate, block->final_rate);
+          }
+          else
+            rate = block->final_rate;
+          return rate * block->xy_to_e_step_ratio;
+        }
+        t -= block->deceleration_time;
+      }
+      return 0.0f;
+    };
+    
+    hal_timer_t Stepper::smooth_lin_adv_isr() {
+      static float soothed_values[SMOOTH_LIN_ADV_EXP_ORDER] = {0};
+      float target_pressure = 0;
+      if (current_block) {
+        bool is_unretracting = !current_block->use_advance_lead && (current_block->direction_bits.e);
+        // Don't lookahead during unretractions to keep pressure zero at start of lines
+        // and so avoid blobs and improve seams
+        if (!is_unretracting) {
+          uint32_t t = Planner::extruder_advance_TAU_TICKS + curr_timer_tick;
+          target_pressure = lookahead(t) * Planner::extruder_advance_K[0];
+        }
+      }
+      else {
+        curr_step_rate = 0;
+      }
+
+      const float last_pressure = soothed_values[SMOOTH_LIN_ADV_EXP_ORDER - 1];
+      for (uint8_t i = 0; i < SMOOTH_LIN_ADV_EXP_ORDER; i++) {
+        // Approximate gaussian smoothing via higher order exponential smoothing
+        target_pressure =
+          Planner::extruder_advance_ALPHA * target_pressure +
+          (1 - Planner::extruder_advance_ALPHA) * soothed_values[i];
+        soothed_values[i] = target_pressure;
+      }
+      const float dt_inv = SMOOTH_LIN_ADV_HZ;
+      const float la_step_rate = (target_pressure - last_pressure) * dt_inv;
+
+      float planned_step_rate = 0;
+      if (current_block) {
+        planned_step_rate = curr_step_rate * current_block->xy_to_e_step_ratio;
+      }
+      float total_rate = planned_step_rate + la_step_rate;
+
+      #if ENABLED(INPUT_SHAPING_E_SYNCH)
+        XYPair pre_shaping_rate = {0, 0},
+               first_pulse_rate = {0, 0};
+        float unshaped_rate_e = total_rate;
+        if (current_block) {
+          float xy_length = current_block->steps.x + current_block->steps.y;
+          if (xy_length > 0) {
+            unshaped_rate_e = 0;
+            pre_shaping_rate = {
+              total_rate * current_block->steps.x / xy_length,
+              total_rate * current_block->steps.y / xy_length,
+            };
+            first_pulse_rate = {
+              pre_shaping_rate.x * Stepper::shaping_x.factor1 / 128.0f,
+              pre_shaping_rate.y * Stepper::shaping_y.factor1 / 128.0f
+            };
+          }
+        }
+        const XYPair second_pulse_rate = {
+          lookback(ShapingQueue::get_delay_x()).x * Stepper::shaping_x.factor2 / 128.0f,
+          lookback(ShapingQueue::get_delay_y()).y * Stepper::shaping_y.factor2 / 128.0f
+        };
+        add_to_buffer(pre_shaping_rate);
+
+        const float x = first_pulse_rate.x + second_pulse_rate.x;
+        const float y = first_pulse_rate.y + second_pulse_rate.y;
+
+        total_rate = unshaped_rate_e + x + y;
+      #endif
```

Comment:
Input shaping can be on any combination of X, Y and Z. Ideally this code should reflect that. (As should the related check `SanityCheck.h`.) However, as remarked at the top of the PR a better solution is probably to synchronise axes in IS, post the LA code.

Reply:
Yes, I avoided doing that because:
* I assume it is unusual not to have IS in both x and y (z may or may not be enabled)
* It also changes the lookback buffer's shape
* It will make the first PR on the topic harder to review
* and _"as remarked at the top of the PR a better solution is probably to synchronise axes in IS, post the LA code"_

Reply:
I am happy with cleaning this up as a separate PR in the IS code.